docs: Remark Plugin tutorial (#13234)
gatsbyjs/gatsby: docs/docs/remark-ast.md

---
title: Manipulating Remark ASTs
---

gatsby-transformer-remark empowers developers to translate Markdown into HTML to be consumed throughout Gatsby pages. Blogs and other content based sites can highly benefit from this so the code and content is separated. With this, people who will write content for the site don't need to worry about how the the site was written, but rather can focus just on writing the Markdown.

> In certain instances, a developer may want to customize certain Markdown nodes while being parsed. Examples could include [add syntax highlighting](/packages/gatsby-remark-prismjs/), [parse images](/packages/gatsby-remark-images), [embed videos](/packages/gatsby-remark-embed-video) and plenty of others. Throughout all of these instances, a plugin will examine the Markdown Abstract Syntax Tree (AST) and manipulate content based on certain node types or content in particular nodes.

Reviewer: ```suggestion
In certain instances, a developer may want to customize the content of the Markdown file and extend it functionally in useful ways. For example, use cases such as [adding syntax highlighting](/packages/gatsby-remark-prismjs/), [parsing and creating responsive images](/packages/gatsby-remark-images), [embedding videos](/packages/gatsby-remark-embed-video), and many others. In each of these examples, a plugin will be injected with the Markdown Abstract Syntax Tree (AST) and manipulate content based on certain node types or content in particular nodes.
```

I wonder if we can improve the flow here by the way. It seems that introducing a perhaps confusing/advanced topic such as ASTs would benefit from an example _early on_. I know we do that later--but just thinking aloud.

Not prescriptive here, but curious what you think!

Author: my mindset here was to give some examples of plugins and then give a mention for how they are done by introducing an AST. @marcysutton do you have any opinion on this?

Reviewer 2: I think keeping it general like Dustin has written makes the paragraph a little more approachable, with lower cognitive overhead before going into more detail. The only change I would make is to ensure a complete sentence:

```suggestion
In certain instances, a developer may want to customize the content of the Markdown file and extend it functionally in useful ways; for example, use cases such as [adding syntax highlighting](/packages/gatsby-remark-prismjs/), [parsing and creating responsive images](/packages/gatsby-remark-images), [embedding videos](/packages/gatsby-remark-embed-video), and many others. In each of these examples, a plugin will be injected with the Markdown Abstract Syntax Tree (AST) and manipulate content based on certain node types or content in particular nodes.
```

Author: Done

## What you will learn in this tutorial

> - How to learn how to understand the remark AST

Reviewer: Weird sentence here:

```suggestion
- Further understanding of the remark Abstract Syntax Tree (AST)
```

> - How to integrate plugins with `gatsby-transformer-remark`

Reviewer: ```suggestion
- How to create a plugin that is injected with an AST via `gatsby-transformer-remark`
```

- How to manipulate the remark AST to add additional functionality

## Understanding the Abstract Syntax Tree

To get an understanding at what is available in the Markdown AST, take a look at the markdown ast spec that is used in remark and other unist projects: [syntax-tree/mdast](https://github.com/syntax-tree/mdast).

Starting out with a Markdown file as below:

```markdown
# Hello World!

This is a [Real page](https://google.com)
```

> Remark would translate this into an AST that will be available to `gatsby-transformer-remark` plugins. [AST Explorer](https://astexplorer.net/#/gist/d9029a2e8827265fbb9b190083b59d4d/3384f3ce6a3084e50043d0c8ce34628ed7477603) is a site that gives you a side-by-side view of the markdown and the outputted AST.

Reviewer: Love the link to AST Explorer! For people who don't click through--it could be worth showing the JSON representation here, as well.

Author: Okay, I'll revert it so both the link and the full JSON is available

Author: Done

## Setting up a plugin

You are going to create a plugin that colors all top-level headings in the markdown with the color purple.

> First create a local plugin by adding a `plugins` folder in your site and generating a package.json file for it. As well, create an index.js file. In this file, it will export a single function.

Reviewer: ```suggestion
First create a local plugin by adding a `plugins` folder in your site and generating a package.json file for it. As well, create an index.js file. In this file, we export a function that will be invoked by `gatsby-transformer-remark`.
```

```js:title=plugins/gatsby-remark-purple-headers/index.js
module.exports = ({ markdownAST }, pluginOptions) => {
  // Manipulate AST

  return markdownAST
}
```

> The first parameter is all of the default properties that can be used in plugins (actions, store, getNodes, schema, etc.) plus a couple just for gatsby-transformer-remark plugins. The one to be focused on is the `markdownAST` field which is destructured in the code snippet above.

Reviewer: ```suggestion
The first parameter is all of the default properties that can be used in plugins (actions, store, getNodes, schema, etc.) plus a couple just for gatsby-transformer-remark plugins. The most relevant field for our purposes is the `markdownAST` field which is destructured in the code snippet above.
```

> As with other Gatsby plugins, the 2nd parameter is the `pluginOptions` which is obtained from the definition in `gatsby-config.js` file.

Reviewer: Show an example here, perhaps?

e.g. given this gatsby-config.js _this_ is the contents of pluginOptions.

Author: Maybe try putting in an aside here so I can have a hex color string in the pluginOptions so the color can be customized? I want it to fit smoothly and not take a tangent.

> Finally, the function will return the `markdownAST` after the fields you wish to be edited are transformed.

Reviewer: Doesn't have to since it's mutated--but a nit ;)

Reviewer: Also I like showing best practices here--so even though we're mutating, returning is a best practice for e.g. easy authoring of unit tests.

## Adding the plugin to your site

> You likely will want to grab `gatsby-source-filesystem` to grab the file nodes initially. In this example it is assumed that the markdown files exist in a `src/data/` directory.

Reviewer: ```suggestion
You likely will want to grab `gatsby-source-filesystem` to inject the file nodes into Gatsby's GraphQL schema. In this example it is assumed that the Markdown files exist in a `src/data/` directory.
```

Be deliberate about casing here, e.g. `markdown` -> `Markdown`

The plugin is now initially set so you can add it as a sub-plugin inside `gatsby-transformer-remark`

```js:title=gatsby-config.js
module.exports = {
  plugins: [
    {
      resolve: "gatsby-source-filesystem",
      options: {
        name: `data`,
        path: `${__dirname}/src/data/`,
      },
    },
    {
      resolve: `gatsby-transformer-remark`,
      options: {
        plugins: [`gatsby-remark-purple-headers`],
      },
    },
  ],
}
```

If you want to add some options, you could switch to the object syntax:

```js
{
  resolve: `gatsby-remark-purple-headers`,
  options: {
    // Options here
  }
}
```

## Find and Modify Markdown Nodes

When modifying nodes, you'll want to walk the tree and then implement new functionality on specific nodes.

> A node module to help with is [unist-util-visit](https://github.com/syntax-tree/unist-util-visit), a walker for `unist` nodes. For reference, Unist which stands for United Syntax Tree is a standard for markdown syntax trees and parsers that include well known parsers in the Gatsby world like Remark and MDX.

Reviewer: ```suggestion
A node module to help with is [unist-util-visit](https://github.com/syntax-tree/unist-util-visit), a walker for `unist` nodes. For reference, Unist (Unified Syntax Tree) is a standard for Markdown syntax trees and parsers that include well known parsers in the Gatsby world like Remark and MDX.
```

As an example from `unist-util-visit`'s readme, it allows for an easy interface to visit particular nodes based on a particular type:

```js
var remark = require("remark")
var visit = require("unist-util-visit")

var tree = remark().parse("Some _emphasis_, **importance**, and `code`.")

visit(tree, "text", visitor)

function visitor(node) {
  console.log(node)
}
```

Here, it finds all text nodes and will `console.log` the nodes.

> Now with this, you can take the AST from your plugin and get working on adding functionality for headers:

Reviewer: ```suggestion
With this technique in mind, you can similarly traverse the AST from your plugin and add additional functionality, like so:
```

```js:title=plugins/gatsby-remark-purple-headers/index.js
const visit = require("unist-util-visit")

module.exports = ({ markdownAST }, pluginOptions) => {

>   visit(markdownAST, "heading", node => {

Reviewer: ```suggestion
  // highlight-next-line
  visit(markdownAST, "heading", node => {
```

        "end": { "line": 1, "column": 15, "offset": 14 },
        "indent": []
      }
    }
  ],
  "position": {
    "start": { "line": 1, "column": 1, "offset": 0 },
    "end": { "line": 1, "column": 15, "offset": 14 },
    "indent": []
  }
},
```

You have context about the text as well as what depth the header is (for instance here you have a depth of 1 which would equate to an `h1` element)

With the inner function of the `visit` call, you parse out all of the text and if it will map to a h1, you set the type of the node to `html` and set the node's value to be some custom HTML.

```js
const visit = require("unist-util-visit")
const toString = require("mdast-util-to-string")

module.exports = ({ markdownAST }, pluginOptions) => {
  visit(markdownAST, "heading", node => {
    let { depth } = node

    // Skip if not an h1
    if (depth !== 1) return

    // Grab the innerText of the heading node
    let text = toString(node)

    const html = `
        <h1 style="color: rebeccapurple">
          ${text}
        </h1>
      `

    node.type = "html"
    node.children = undefined
    node.value = html
  })

  return markdownAST
}
```

> A small library [mdast-util-to-string](https://github.com/syntax-tree/mdast-util-to-string) which is also made by Unified was used to extract the plaintext of the inner nodes, so this would remove links or other types of nodes inside the header, but given you have full access to the markdown AST, you can modify it however you wish.

Reviewer: ```suggestion
A small library [mdast-util-to-string](https://github.com/syntax-tree/mdast-util-to-string) which is provided by Unified was used to extract the plain text of the inner nodes, so this would remove links or other types of nodes inside the header, but given you have full access to the markdown AST, you can modify it however you wish.
```

Also would probably tweak this sentence--it's a bit of a doozy 😩

Reviewer 2: I think it could be even shorter/broken up:
```suggestion
A small library [mdast-util-to-string](https://github.com/syntax-tree/mdast-util-to-string) by Unified was used to extract the plain text of the inner nodes. This would remove links or other types of nodes inside the heading, but given you have full access to the markdown AST, you can modify it however you wish.
```

## Loading in changes and seeing effect

> After doing such, you setup some pages to be programmatically created from markdown as shown in [Part 7 of the Gatsby Tutorial](/tutorial/part-seven/). Once this is set up, you can examine that your plugin works as seen below based on the markdown you wrote earlier.

Reviewer: ```suggestion
At this point, our plugin is now ready to be used. To see the resulting functionality, it is helpful to re-visit [Part 7 of the Gatsby Tutorial](/tutorial/part-seven/) to programmatically create pages from Markdown data. Once this is set up, you can examine that your plugin works as seen below based on the markdown you wrote earlier.
```

Hmm... bit of a jump here. Perhaps we can add this as a pre-requisite or mention this earlier?

Author: Maybe I could add an additional section at the top of prerequisites at the top listing things that you should have under your belt such as going through the Gatsby Tutorial.

Author: Took the suggestion and added a prerequisite section

---
title: Manipulating Remark ASTs
---

> [`gatsby-transformer-remark`](/plugins/gatsby-transformer-remark) empowers developers to translate Markdown into HTML to be consumed via Gatsby's GraphQL API. Blogs and other content based sites can highly benefit from functionality enabled with this plugin. With this plugin, authors of content for the site don't need to worry about how the the site is written or structured but can rather focus on writing engaging posts and content!

Reviewer 2: ```suggestion
[`gatsby-transformer-remark`](/plugins/gatsby-transformer-remark) empowers developers to translate Markdown into HTML to be consumed via Gatsby's GraphQL API. Blogs and other content based sites can highly benefit from functionality enabled with this plugin. With this plugin, authors of content for the site don't need to worry about how the site is written or structured but can rather focus on writing engaging posts and content!
```

> In certain instances, a developer may want to customize certain Markdown nodes while being parsed. Examples could include [add syntax highlighting](/packages/gatsby-remark-prismjs/), [parse images](/packages/gatsby-remark-images), [embed videos](/packages/gatsby-remark-embed-video) and plenty of others. Throughout all of these instances, a plugin will examine the Markdown Abstract Syntax Tree (AST) and manipulate content based on certain node types or content in particular nodes.

Reviewer 2: ```suggestion
In certain instances, a developer may want to customize certain Markdown nodes while being parsed. Examples could include [adding syntax highlighting](/packages/gatsby-remark-prismjs/), [parse images](/packages/gatsby-remark-images), [embedding videos](/packages/gatsby-remark-embed-video) and plenty of others. Throughout all of these instances, a plugin will examine the Markdown Abstract Syntax Tree (AST) and manipulate content based on certain node types or content in particular nodes.
```

## What you will learn in this tutorial

- Further understanding of the remark Abstract Syntax Tree (AST)
- How to create a plugin that is injected with an AST via `gatsby-transformer-remark`
- How to manipulate the remark AST to add additional functionality

## Understanding the Abstract Syntax Tree

> To get an understanding at what is available in the Markdown AST, take a look at the markdown ast spec that is used in remark and other unist projects: [syntax-tree/mdast](https://github.com/syntax-tree/mdast).

Reviewer 2: ```suggestion
To get an understanding at what is available in the Markdown AST, take a look at the Markdown AST spec that is used in remark and other unist projects: [syntax-tree/mdast](https://github.com/syntax-tree/mdast).
```
Also, what is `unist`?

Starting out with a Markdown file as below:

```markdown
# Hello World!

This is a [Real page](https://google.com)
```

> Remark would translate this into an AST that will be available to `gatsby-transformer-remark` plugins. [AST Explorer](https://astexplorer.net/#/gist/d9029a2e8827265fbb9b190083b59d4d/3384f3ce6a3084e50043d0c8ce34628ed7477603) is a site that gives you a side-by-side view of the markdown and the outputted AST.

Reviewer 2: ```suggestion
Remark would translate this into an AST made available to `gatsby-transformer-remark` plugins. [AST Explorer](https://astexplorer.net/#/gist/d9029a2e8827265fbb9b190083b59d4d/3384f3ce6a3084e50043d0c8ce34628ed7477603) is a site that gives you a side-by-side view of the markdown and the outputted AST.
```

## Setting up a plugin

You are going to create a plugin that colors all top-level headings in the markdown with the color purple.

> First create a local plugin by adding a `plugins` folder in your site and generating a package.json file for it. As well, create an index.js file. In this file, we export a function that will be invoked by `gatsby-transformer-remark`.

Reviewer 2: ```suggestion
First create a local plugin by adding a `plugins` folder in your site and generating a `package.json` file for it. As well, create an `index.js` file. In this file, we export a function that will be invoked by `gatsby-transformer-remark`:
```

The plugin is now initially set so you can add it as a sub-plugin inside `gatsby-transformer-remark`

```js:title=gatsby-config.js
module.exports = {
  plugins: [
    {
      resolve: "gatsby-source-filesystem",
      options: {
        name: `data`,
        path: `${__dirname}/src/data/`,
      },
    },
    {
      resolve: `gatsby-transformer-remark`,
      options: {
        plugins: [`gatsby-remark-purple-headers`],
      },
    },
  ],
}
```

If you want to add some options, you could switch to the object syntax:

```js
{
  resolve: `gatsby-remark-purple-headers`,
  options: {
    // Options here
  }
}
```

## Find and Modify Markdown Nodes

When modifying nodes, you'll want to walk the tree and then implement new functionality on specific nodes.

A node module to help with is [unist-util-visit](https://github.com/syntax-tree/unist-util-visit), a walker for `unist` nodes. For reference, Unist (Unified Syntax Tree) is a standard for Markdown syntax trees and parsers that include well known parsers in the Gatsby world like Remark and MDX.

> As an example from `unist-util-visit`'s readme, it allows for an easy interface to visit particular nodes based on a particular type:

Reviewer 2: ```suggestion
As an example from `unist-util-visit`'s README file, it allows for an easy interface to visit particular nodes based on a particular type:
```

```js
var remark = require("remark")
var visit = require("unist-util-visit")

var tree = remark().parse("Some _emphasis_, **importance**, and `code`.")

visit(tree, "text", visitor)

function visitor(node) {
  console.log(node)
}
```

Here, it finds all text nodes and will `console.log` the nodes.

With this technique in mind, you can similarly traverse the AST from your plugin and add additional functionality, like so:

```js:title=plugins/gatsby-remark-purple-headers/index.js
const visit = require("unist-util-visit")

module.exports = ({ markdownAST }, pluginOptions) => {
  // highlight-next-line
  visit(markdownAST, "heading", node => {
    // Do stuff with heading nodes
  })

  return markdownAST
}
```

> Next, by visiting all heading nodes and passing them into a transformer function, you can manipulate the particular nodes to match your usecase.

Reviewer 2: ```suggestion
Next, by visiting all heading nodes and passing them into a transformer function, you can manipulate the particular nodes to match your use case.
```

> Looking again at the AST node for header:

Reviewer 2: ```suggestion
Looking again at the AST node for heading:
```
Looking at the JSON structure and nature of the tutorial, "heading" is probably more accurate than "header" in an HTML / markup context.

```JSON
{
  "type": "heading",
  "depth": 1,
  "children": [
    {
      "type": "text",
      "value": "Hello World!",
      "position": {
        "start": { "line": 1, "column": 3, "offset": 2 },
        "end": { "line": 1, "column": 15, "offset": 14 },
        "indent": []
      }
    }
  ],
  "position": {
    "start": { "line": 1, "column": 1, "offset": 0 },
    "end": { "line": 1, "column": 15, "offset": 14 },
    "indent": []
  }
},
```

> You have context about the text as well as what depth the header is (for instance here you have a depth of 1 which would equate to an `h1` element)

Reviewer 2: ```suggestion
You have context about the text as well as what depth the heading is (for instance here you have a depth of 1 which would equate to an `h1` element)
```

    const html = `
        <h1 style="color: rebeccapurple">
          ${text}
        </h1>
      `

    node.type = "html"
    node.children = undefined
    node.value = html
  })

  return markdownAST
}
```

A small library [mdast-util-to-string](https://github.com/syntax-tree/mdast-util-to-string) which is also made by Unified was used to extract the plaintext of the inner nodes, so this would remove links or other types of nodes inside the header, but given you have full access to the markdown AST, you can modify it however you wish.

## Loading in changes and seeing effect

After doing such, you setup some pages to be programmatically created from markdown as shown in [Part 7 of the Gatsby Tutorial](/tutorial/part-seven/). Once this is set up, you can examine that your plugin works as seen below based on the markdown you wrote earlier.

![Output](images/remark-ast-output.png)

## Publishing the plugin

To be shared with others, you can extract the plugin to its own directory outside of this site and then publish it to NPM so it can be accessed both on NPM and [Submitted to the Plugin Library](/contributing/submit-to-plugin-library).

## Summary

> You just wrote a local Gatsby plugin that is a sub-plugin for `gatsby-transformer-remark` that Manipulates the Remark AST and have a further understanding about the structure of Markdown Abstract Syntax Trees

Reviewer 2: ```suggestion
You just wrote a local Gatsby plugin that is a sub-plugin for `gatsby-transformer-remark` that manipulates the Remark AST. You should now have a further understanding about the structure of Markdown Abstract Syntax Trees! Yay!
```
We should congratulate people for making it this far!

---
title: Manipulating Remark ASTs
---

> gatsby-transformer-remark empowers developers to translate markdown into html to be consumed throughout Gatsby pages. Blogs and other content based sites can highly benefit from this so the code and content is separated and creators don't need.

Author: Capitalized Markdown and HTML, as well as finished the end of the paragraph that creators don't need to focus on the code and can just be concerned with the Markdown content.

In certain instances, a developer may want to customize certain markdown nodes while being parsed. Examples could include [add syntax highlighting](/packages/gatsby-remark-prismjs/), [parse images](/packages/gatsby-remark-images), [embed videos](/packages/gatsby-remark-embed-video) and plenty of others. Throughout all of these instances, a plugin will examine the markdown Abstract Syntax Tree (AST) and manipulate content based on certain node types or content in particular nodes.

> ## What will be learned in this tutorial

Reviewer 3: ```suggestion
## What you will learn in this tutorial
```

Active voice is better!

- How to learn how to understand the remark AST
- How to integrate plugins with `gatsby-transformer-remark`
- How to manipulate the remark AST to add additional functionality.

> ## Understanding the Syntax Tree

Reviewer 3: ```suggestion
## Understanding the Abstract Syntax Tree
```

          "value": "This is a ",
          "position": {
            "start": { "line": 3, "column": 1, "offset": 16 },
            "end": { "line": 3, "column": 11, "offset": 26 },
            "indent": []
          }
        },
        {
          "type": "link",
          "title": null,
          "url": "https://google.com",
          "children": [
            {
              "type": "text",
              "value": "Real page",
              "position": {
                "start": { "line": 3, "column": 12, "offset": 27 },
                "end": { "line": 3, "column": 21, "offset": 36 },
                "indent": []
              }
            }
          ],
          "position": {
            "start": { "line": 3, "column": 11, "offset": 26 },
            "end": { "line": 3, "column": 42, "offset": 57 },
            "indent": []
          }
        }
      ],
      "position": {
        "start": { "line": 3, "column": 1, "offset": 16 },
        "end": { "line": 3, "column": 42, "offset": 57 },
        "indent": []
      }
    }
  ],
  "position": {
    "start": { "line": 1, "column": 1, "offset": 0 },
    "end": { "line": 4, "column": 1, "offset": 58 }
  }
}
```

> This is the syntax tree that will be parsed by remark into HTML by `gatsby-transformer-remark`.

Reviewer 3: How about adding a link to [AST explorer](https://astexplorer.net/#/gist/d9029a2e8827265fbb9b190083b59d4d/3384f3ce6a3084e50043d0c8ce34628ed7477603) here?

Author: Removed the large AST block and linked to the AST explorer as provided

## Setting up a plugin

> For example, we are going to create a plugin that colors all top-level headings in the markdown with the color purple.

Reviewer 3: ```suggestion
You are going to create a plugin that colors all top-level headings in the markdown with the color purple.
```

First create a local plugin by adding a `plugins` folder in your site and generating a package.json file for it. As well, create an index.js file. In this file, it will export a single function.

```js:title=plugins/gatsby-remark-purple-headers/index.js
module.exports = ({ markdownAST }, pluginOptions) => {
  // Manipulate AST

  return markdownAST
}
```

The first parameter is all of the default properties that can be used in plugins (actions, store, getNodes, schema, etc.) plus a couple just for gatsby-transformer-remark plugins. The one to be focused on is the `markdownAST` field which is destructured in the code snippet above.

> As with other Gatsby plugins, the 2nd parameter is the pluginOptions which is obtained from the definition in `gatbsy-config.js` file.

Reviewer 3: ```suggestion
As with other Gatsby plugins, the 2nd parameter is the `pluginOptions` which is obtained from the definition in `gatsby-config.js` file.
```

> Finally, the function will return the markdownAST after the fields you wish to be edited are transformed.

Reviewer 3: ```suggestion
Finally, the function will return the `markdownAST` after the fields you wish to be edited are transformed.
```

## Adding the plugin to your site

> You likely will want to grab `gatsby-source-filesystem` to grab the file nodes initially. in this example I am going to assume the markdown files we are working with exist in a `src/data/` directory.

Reviewer 3: ```suggestion
You likely will want to grab `gatsby-source-filesystem` to grab the file nodes initially. In this example it is assumed that the markdown files exist in a `src/data/` directory.
```

> The plugin is now initially set so we can add it as a sub-plugin inside `gatsby-transformer-remark`

Reviewer 3: ```suggestion
The plugin is now initially set so you can add it as a sub-plugin inside `gatsby-transformer-remark`
```

you instead of we :)

```js:title=gatsby-config.js
module.exports = {
  plugins: [
    {
      resolve: "gatsby-source-filesystem",
      options: {
        name: `data`,
        path: `${__dirname}/src/data/`,
      },
    },
    {
      resolve: `gatsby-transformer-remark`,
      options: {
        plugins: [`gatsby-remark-purple-headers`],
      },
    },
  ],
}
```

> If we wanted to add some options, we could switch to the object syntax:

Reviewer 3: ```suggestion
If you want to add some options, you could switch to the object syntax:
```

```js
{
  resolve: `gatsby-remark-purple-headers`,
  options: {
    // Options here
  }
}
```

## Find and Modify Markdown Nodes

When modifying nodes, you'll want to walk the tree and then implement new functionality on specific nodes.

A node module to help with is [unist-util-visit](https://github.com/syntax-tree/unist-util-visit), a walker for `unist` nodes. For reference, Unist which stands for United Syntax Tree is a standard for markdown syntax trees and parsers that include well known parsers in the Gatsby world like Remark and MDX.

As an example from `unist-util-visit`'s readme, it allows for an easy interface to visit particular nodes based on a particular type:

```js
var remark = require("remark")
var visit = require("unist-util-visit")

var tree = remark().parse("Some _emphasis_, **importance**, and `code`.")

visit(tree, "text", visitor)

function visitor(node) {
  console.log(node)
}
```

Here, it finds all text nodes and will `console.log` the nodes.

> Now with this, we can take the AST from our plugin and get working on adding functionality for headers:

Reviewer 3: ```suggestion
Now with this, you can take the AST from your plugin and get working on adding functionality for headers:
```

> ```js

Reviewer 3: ```suggestion
```js:title=plugins/gatsby-remark-purple-headers/index.js
```

const visit = require("unist-util-visit")

module.exports = ({ markdownAST }, pluginOptions) => {
  visit(markdownAST, "heading", node => {
    // Do stuff with heading nodes
  })

  return markdownAST
}
```

> Now we will visit all heading nodes and pass it into a transformer function which we can do whatever we wish.

Reviewer 3: ```suggestion
Now you will visit all heading nodes and pass it into a transformer function which you can do whatever we wish.
```

I can't come up with something better myself right now, but the "Now you will" should probably be replaced with some other introductions. You can probably remove it completely and just tell what to do, not what one will do now.

Author: Rephrased this paragraph. Hopefully it works better.

Looking again at the AST node for header:

```JSON
{
  "type": "heading",
  "depth": 1,
  "children": [
    {
      "type": "text",
      "value": "Hello World!",
      "position": {
        "start": { "line": 1, "column": 3, "offset": 2 },
        "end": { "line": 1, "column": 15, "offset": 14 },
        "indent": []
      }
    }
  ],
  "position": {
    "start": { "line": 1, "column": 1, "offset": 0 },
    "end": { "line": 1, "column": 15, "offset": 14 },
    "indent": []
  }
},
```

> We have context about the text as well as what depth the header is (for instance here we have a depth of 1 which would equate to an `h1` element)

Reviewer 3: ```suggestion
You have context about the text as well as what depth the header is (for instance here you have a depth of 1 which would equate to an `h1` element)
```

> with the inner function of the `visit` call, I parse out all of the text and if it will map to a h1, I set the type of the node to `html` and set the node's value to be some custom html.

Reviewer 3: ```suggestion
With the inner function of the `visit` call, you parse out all of the text and if it will map to a h1, you set the type of the node to `html` and set the node's value to be some custom HTML.
```

> ```js

Reviewer 3: Isn't this code snippet missing some content?

Author: Added the remaining code around the code snippet to have the full example here

const toString = require("mdast-util-to-string")

visit(markdownAST, "heading", node => {
  let { depth } = node

  // Skip if not an h1
  if (depth !== 1) return

  // Grab the innerText of the heading node
  let text = toString(node)

  const html = `
      <h1 style="color: rebeccapurple">
        ${text}
      </h1>
    `

  node.type = "html"
  node.children = undefined
  node.value = html
})
```

A small library [mdast-util-to-string](https://github.com/syntax-tree/mdast-util-to-string) which is also made by Unified was used to extract the plaintext of the inner nodes, so this would remove links or other types of nodes inside the header, but given you have full access to the markdown AST, you can modify it however you wish.

> ## Loading in changes and seeing effect.

Reviewer 3: ```suggestion
## Loading in changes and seeing effect
```

> After doing such, we setup some pages to be programmatically created from markdown as shown in [Part 7 of the Gatsby Tutorial](/tutorial/part-seven/). Once this is set up, we can examine that our plugin works as seen below based on the markdown we wrote earlier.

Reviewer 3: ```suggestion
After doing such, you setup some pages to be programmatically created from markdown as shown in [Part 7 of the Gatsby Tutorial](/tutorial/part-seven/). Once this is set up, you can examine that your plugin works as seen below based on the markdown you wrote earlier.
```

![Output](images/remark-ast-output.png)

## Publishing the plugin

> To be shared with others, we can extract the plugin to its own directory outside of this site and then publish it to NPM so it can be accessed both on NPM and [Submitted to the Plugin Library](/contributing/submit-to-plugin-library).

Reviewer 3: ```suggestion
To be shared with others, you can extract the plugin to its own directory outside of this site and then publish it to NPM so it can be accessed both on NPM and [Submitted to the Plugin Library](/contributing/submit-to-plugin-library).
```

## Summary

You just wrote a local Gatsby plugin that is a sub-plugin for `gatsby-transformer-remark` that Manipulates the Remark AST and have a further understanding about the structure of Markdown Abstract Syntax Trees

> ## Where next?

Reviewer 3: ```suggestion
## What's next?
```
